spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -1145,45 +1161,62 @@ def write_units(
+
             for unit_id in unit_ids:
-                unit_kwargs = dict()
-                # spike trains withinin the SortingExtractor object are not scaled by sampling frequency
+                unit_kwargs = dict(id=unit_id)
+
+                # spike trains withinin the SortingExtractor object are in units of sampling frames
                 if timestamps is not None:
                     spike_train_frames = sorting.get_unit_spike_train(unit_id=unit_id)
-                    assert spike_train_frames[-1] < len(timestamps), "Number of 'timestamps' differs from number of " \
-                                                                     "'frames'!"
-                    spkt = np.array(timestamps)[spike_train_frames]
+                    assert spike_train_frames[-1] < len(timestamps), \
+                        "Number of 'timestamps' differs from number of 'frames'!"
+                    unit_kwargs.update(spike_times=np.array(timestamps)[spike_train_frames])
                 else:
-                    spkt = sorting.get_unit_spike_train(unit_id=unit_id) / fs
+                    unit_kwargs.update(spike_times=sorting.get_unit_spike_train(unit_id=unit_id) / sf)
+
                 for pr in write_properties:
                     if pr in sorting.get_unit_property_names(unit_id):
                         prop_value = sorting.get_unit_property(unit_id, pr)
                         unit_kwargs.update({pr: prop_value})
                     else:  # Case of missing data for this unit and this property
                         unit_kwargs.update({pr: np.nan})
-                nwbfile.add_unit(id=unit_id, spike_times=spkt, **unit_kwargs)
 
-            # TODO
-            # # Stores average and std of spike traces
-            # This will soon be updated to the current NWB standard
-            # if 'waveforms' in sorting.get_unit_spike_feature_names(unit_id=id):
-            #     wf = sorting.get_unit_spike_features(unit_id=id,
-            #                                          feature_name='waveforms')
-            #     relevant_ch = most_relevant_ch(wf)
-            #     # Spike traces on the most relevant channel
-            #     traces = wf[:, relevant_ch, :]
-            #     traces_avg = np.mean(traces, axis=0)
-            #     traces_std = np.std(traces, axis=0)
-            #     nwbfile.add_unit(
-            #         id=id,
-            #         spike_times=spkt,
-            #         waveform_mean=traces_avg,
-            #         waveform_sd=traces_std
-            #     )
+                if write_waveforms:
+                    check_waveform_features(sorting=sorting, unit_id=unit_id)
+                    wf = sorting.get_unit_spike_features(unit_id=unit_id, feature_name='waveforms')
+                    # This is temporary until the NWB Schema allows conversion factors for waveforms in the units table
+                    wf = wf * gain * 1e-6
```

Comment:
why not leaving them in uV?

Reply:
Unfortunately, when we added waveforms to the units table in the NWB schema we didn't add a field for a conversion factor that would allow us to store the base dtype separately. This is on the to do list, but schema related things tend to take a long time to get all the way through the pipeline.